core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -847,6 +847,305 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::TeamThreadRangeBoundariesStruct<
+//     - total_rank is the total number of loop nests
+//     - iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - ParThreadRank is the nesting level on which to deploy thread parallelism
+//   - ParVectorRank is the nesting level on which to deploy vector parallelism
+//   - CurrentRank is the current nest level
+template <typename Rank, int ParThreadRank, int ParVectorRank, int CurrentRank>
+struct TeamMDRangeRanks {
+  using RankType                            = int;
+  static constexpr Iterate iter             = Rank::outer_direction;
+  static constexpr RankType total_rank      = Rank::rank;
+  static constexpr RankType par_thread_rank = ParThreadRank;
+  static constexpr RankType par_vector_rank = ParVectorRank;
+  static constexpr RankType current_rank    = CurrentRank;
+
+  // We have to recursively process ranks [0..total_rank-1]
+  using RangeMode = TeamMDRangeMode<
+      (iter == Iterate::Right)
+          ? static_cast<TeamMDRangeEndRank>(current_rank == total_rank)
+          : static_cast<TeamMDRangeEndRank>(current_rank == -1),
+      static_cast<TeamMDRangeParThread>(current_rank == par_thread_rank),
+      static_cast<TeamMDRangeParVector>(current_rank == par_vector_rank)>;
+};
+
+// ParallelRank determines on which rank parallelization happens.
+//   - Rank is Kokkos::Rank<TotalRank, Iter>
+//     - TotalRank is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - ThreadAndVector determines whether both vector and thread parallelism is
+//   in use
+template <typename Rank, typename ExecSpace,
+          TeamMDRangeThreadAndVector ThreadAndVector>
+struct ParallelRank;
+
+#ifdef KOKKOS_ENABLE_SERIAL
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ParallelRank<Rank, Serial, ThreadAndVector> {
+  static constexpr bool is_direction_left =
+      (Rank::outer_direction == Iterate::Left);
+  static constexpr int par_rt  = is_direction_left ? Rank::rank - 1 : 0;
+  static constexpr int par_rv  = is_direction_left ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_CUDA
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ParallelRank<Rank, Cuda, ThreadAndVector> {
+  static constexpr bool is_direction_left =
+      (Rank::outer_direction == Iterate::Left);
+  // If vector parallelism is in use deploy thread parallelism on
+  // the second to the last rank, otherwise, thread parallelism on the last rank
+  static constexpr int par_rt =
+      is_direction_left ? ((ThreadAndVector) ? 1 : 0)
+                        : ((ThreadAndVector) ? Rank::rank - 2 : Rank::rank - 1);
+  // Vector parallelism will always be on the last index
+  static constexpr int par_rv  = is_direction_left ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
```

Comment:
O think it might be good to unify here in the categories we got. Essentially GPU vs CPU and do I actually got two nesting levels or not. So HIP and CUDA would do the same OpenMPTarget and SYCL for now do the same, and the CPU backends do the same.

Reply:
Partially handled in [e400aa1](https://github.com/kokkos/kokkos/pull/5238/commits/e400aa1fc708dbc461c9f646041c10e82d30d2be).
Looking for more fitting names for these "strategy" structs.
I will add in another "strategy" struct for OpenMPTarget and SYCL as applicable.